### PR TITLE
Turn on transfer floating point round-off decisions diagnosis by default

### PR DESCRIPTION
**Pull request by GiudGiud**
refs #24226

This is going to have quite a few failures.
I ll try to tell the users to turn it off if not needed

### CODE REVIEW

**Comment by roystgnr** [comment]
Do you have an approximate notion of how expensive it is to leave this on?

Is there a risk of this biting people in moving mesh simulations?  Like, they do a few timesteps, they're comfortable that they haven't got any transfer ambiguity, then they start a whole run and the next thing you know timestep 200 dies because a deforming mesh put a node into the wrong place.  I'm guessing "no", both because deformation will never go *exactly* to the wrong place and because "just turn off the flag and restart afterward" isn't too heinous a workaround; just trying to brainstorm possible failure cases.

**Comment by GiudGiud** [comment]
> Do you have an approximate notion of how expensive it is to leave this on?

Nope. This is equivalent to the greedy search for transfers. Scales as N^2 mpi procs

> Is there a risk of this biting people in moving mesh simulations? Like, they do a few timesteps, they're comfortable that they haven't got any transfer ambiguity, then they start a whole run and the next thing you know timestep 200 dies because a deforming mesh put a node into the wrong place.

Depends what you mean by biting. It s just a warning so it wont crash their simulation. 
If they turn it off, then yes they can be hit by getting a FP round-off determined value. But then again they turned it off

**Comment by GiudGiud** [comment]
Should be mergable independent from the two other PRs btw

**Comment by GiudGiud** [comment]
Seems there's some test failures to investigate in sweeping. Probably more value conflicts but need to make sure

**Comment by GiudGiud** [comment]
Should be green now.

**Comment by GiudGiud** [comment]
These have got to be false positives. I don’t think I was that bad at designing my tests..

**Comment by GiudGiud** [comment]
ok found the problem. It s a false negative that appears when:
- there is an indetermination in what is being sent by a process (think rank 3 holds 2 points that are equidistant from any target point and the closest in rank 3 to that point)
- that value is not actually the "nearest" so it doesnt matter (those 2 points are far from the target point, or simply not as close a point in rank 1)

**Comment by GiudGiud** [comment]
of course the problem only shows itself in nearest node, the most awful one. The other ones dont allow for this nonsense

**Comment by GiudGiud** [comment]
Last commit explains some testing issues. I dont think I m breaking things with this PR, but likely displaying an area of improvement for nearest node as evidenced by some parallel diffs with greedy search

File: framework/include/transfers/MultiAppGeneralFieldTransfer.h [hunk 1]
```diff
@@ -320,15 +320,41 @@ class MultiAppGeneralFieldTransfer : public MultiAppConservativeTransfer
...
+   * distances (used for nodal-value-dof-only variables and constant monomials)
+   * @param distance_caches a cache holding the distances received (used for higher order elemental
+   * variables)
+   * @param conflicts_vec a vector holding all the potential local conflicts
+   */
+  void examineLocalValueConflicts(
+      const unsigned int var_index,
       const DofobjectToInterpValVec & dofobject_to_valsvec,
       const InterpCaches & distance_caches,
-      std::vector<std::tuple<unsigned int, dof_id_type, Point, Real>> conflicts_vec);
```
Comment (roystgnr) [inline framework/include/transfers/MultiAppGeneralFieldTransfer.h:328]: This was an atrocity.  We were accidentally passing by value before, so discarding changes entirely?

File: framework/src/transfers/MultiAppGeneralFieldTransfer.C [hunk 2]
```diff
@@ -743,26 +744,25 @@ MultiAppGeneralFieldTransfer::examineValueConflicts(
...
+    // Extract info for the potential conflict
     const unsigned int problem_id = std::get<0>(potential_conflict);
     const dof_id_type dof_object_id = std::get<1>(potential_conflict);
     const Point p = std::get<2>(potential_conflict);
     const Real distance = std::get<3>(potential_conflict);
 
-    // Extract variable info
+    // Extract target variable info
     auto & es = getEquationSystem(*_to_problems[problem_id], _displaced_target_mesh);
     System * to_sys = find_sys(es, var_name);
```
Comment (roystgnr) [inline framework/src/transfers/MultiAppGeneralFieldTransfer.C:755]: Would it be hard to just save this `System *`, the same way we save the var name now?  Going from a perfectly good index to a string makes me worry about performance (string comparisons don't belong in inner loops) and worry more about robustness (now that we can support multiple systems it's easy to imagine use cases in which we'd want them to have overlapping variable names...)

File: framework/include/transfers/MultiAppGeneralFieldTransfer.h (hunk 1 above)
Comment (roystgnr) [reply framework/include/transfers/MultiAppGeneralFieldTransfer.h:328]: The more I think about it ... does this need to be a separate parameter, even?  We only ever call it with `this->_local_conflicts`, right?  Likewise with `this->_received_conflicts` in that version above?  And this is a non-const non-static member function so we already have access to those that way.

Comment (GiudGiud) [reply framework/include/transfers/MultiAppGeneralFieldTransfer.h:328]: I can remove it from the parameters. Initially I was trying to use the same routine both local and received conflicts but that didn’t last

Comment (roystgnr) [reply framework/include/transfers/MultiAppGeneralFieldTransfer.h:328]: I'm fine with it as-is, if you don't want to change it.  Almost done with the rest of the review, and so far I haven't seen anything that actually *needs* a change.

File: framework/src/transfers/MultiAppGeneralFieldTransfer.C (hunk 2 above)
Comment (GiudGiud) [reply framework/src/transfers/MultiAppGeneralFieldTransfer.C:755]: @friedmud brought the same concern (in the first general field PR) as it's done all over with the names (in fact saving the system does not help, you need to save the variable number as that's retrieved with string comparisons too)

and I profiled and it was irrelevant

overlapping variable names sounds like a bad idea

**Comment by GiudGiud** [comment]
@roystgnr

File: modules/doc/content/newsletter/2023/2023_02.md [hunk 3]
```diff
@@ -9,7 +9,7 @@ based on libMesh's [generic projector](https://mooseframework.inl.gov/docs/doxyg
 explains all the features (siblings transfers, spatial restrictions, higher order variables, etc.) which are supported. The specific instances implemented are:
 
 - [MultiAppGeneralFieldShapeEvaluationTransfer.md]
-- [MultiAppGeneralFieldNearestNodeTransfer.md]
+- [MultiAppGeneralFieldNearestLocationTransfer.md]
```
Comment (cticenhour) [inline modules/doc/content/newsletter/2023/2023_02.md:12]: Thanks for updating these old newsletters. Also thanks for all the transitions to parent/child in the in-file doco and elsewhere.

Comment (GiudGiud) [reply modules/doc/content/newsletter/2023/2023_02.md:12]: you're welcome!

**Comment by GiudGiud** [comment]
so green

**Comment by cticenhour** [comment]
@roystgnr any more comments/review on this PR? I am happy to finish off the review, but I am not as familiar with the current transfers system.

**Review by roystgnr — APPROVED** [review]

**Comment by roystgnr** [comment]
Man, that was so much easier than 25087.  :-)  Looks good to me now!

**Comment by GiudGiud** [comment]
;)

thanks for the review! Last change to transfers hopefully for a while! Hopefully no other false positives to address with this new check

**Comment by aprilnovak** [comment]
FYI this is causing problems with moving mesh cases in Cardinal (see: https://github.com/neams-th-coe/cardinal/pull/788). Sending displacements into Nek, our mesh elements are now giving negative Jacobians. Sorry for being so late to the discussion here, it took some time to do the git bisection search to identify the cause, and we only update MOOSE ~once a month.

Screen Shot 2023-10-07 at 10 16 38 AM

My two cents is that this flag would be better defaulting to "off," if it takes additional time to evaluate & has potential to break moving mesh cases.

**Comment by GiudGiud** [comment]
We pushed a fix recently (last week)

Which commit are you using?

**Comment by aprilnovak** [comment]
Thanks for the tip! I was using a commit just a tad older than that. The hotfix solves our issue, thanks so much.

**Pull request merged**